File: src/main/java/komu/blunt/core/CoreIfExpression.java
package komu.blunt.core;

import komu.blunt.asm.*;

import static com.google.common.base.Preconditions.checkNotNull;

public final class CoreIfExpression extends CoreExpression {

    private final CoreExpression condition;
    private final CoreExpression consequent;
    private final CoreExpression alternative;

    public CoreIfExpression(CoreExpression condition, CoreExpression consequent, CoreExpression alternative) {
        this.condition = checkNotNull(condition);
        this.consequent = checkNotNull(consequent);
        this.alternative = checkNotNull(alternative);
    }

    @Override
    public void assemble(Assembler asm, Instructions instructions, Register target, Linkage linkage) {
        Label after = asm.newLabel("if-after");
        Label falseBranch = asm.newLabel("if-false");

        Linkage trueLinkage = (linkage == Linkage.NEXT) ? Linkage.jump(after) : linkage;

        // Since the target register is safe to overwrite, we borrow it
        // for evaluating the condition as well.
        condition.assemble(asm, instructions, target, Linkage.NEXT);
        instructions.jumpIfFalse(target, falseBranch);

        consequent.assemble(asm, instructions, target, trueLinkage);
        instructions.label(falseBranch);
        alternative.assemble(asm, instructions, target, linkage);
        instructions.label(after);

        instructions.finishWithLinkage(linkage);
    }

    @Override
    public String toString() {
        return "(if " + condition + " " + consequent + " " + alternative + ")";
    }
}

File: src/main/kotlin/komu/blunt/core/CoreIfExpression.kt
package komu.blunt.core

import komu.blunt.asm.*
import komu.blunt.asm.opcodes.OpJumpIfFalse

class CoreIfExpression(val condition: CoreExpression,
                       val consequent: CoreExpression,
                       val alternative: CoreExpression) : CoreExpression() {

    override fun assemble(asm: Assembler, target: Register, linkage: Linkage) =
        instructions {
            val after = asm.newLabel("if-after")
            val falseBranch = asm.newLabel("if-false")

            val trueLinkage = if (linkage == Linkage.Next) Linkage.Jump(after) else linkage

            // Since the target register is safe to overwrite, we borrow it
            // for evaluating the condition as well.

            instructionsOf(condition.assemble(asm, target, Linkage.Next))

            jumpIfFalse(target, falseBranch)

            instructionsOf(consequent.assemble(asm, target, trueLinkage))
            label(falseBranch)
            instructionsOf(alternative.assemble(asm, target, linkage))
            label(after)

            finishWithLinkage(linkage)
        }

    override fun simplify(): CoreExpression {
        val test = condition.simplify()
        val con = consequent.simplify()
        val alt = alternative.simplify()

        return when {
            test is CoreConstantExpression -> if (OpJumpIfFalse.isFalse(test.value)) alt else con
            con == alt                     -> CoreExpression.sequence(test, con)
            else                           -> CoreIfExpression(test, con, alt)
        }
    }

    override fun toString() = "(if $condition $consequent $alternative)"
}
